src/gtk/toga_gtk/widgets/table.py

```diff
@@ -23,7 +25,24 @@ def create(self):
         self.native.interface = self.interface
 
     def refresh(self):
-        raise NotImplementedError()
+        self.data.clear()
+
+        for row in self.interface.data.rows:
+            self.data.append(row.data)
 
     def set_on_select(self, handler):
-        pass
+        for conn_id in self._connections:
+            # Disconnect all other on_select handlers, so that if you reassign
+            # the on_select, it doesn't trigger the old ones too.
+            self.table.disconnect(conn_id)
+
+        def _handler(widget, *args):
+            selection = widget.get_selection()
+            selection.set_mode(Gtk.SelectionMode.SINGLE)
+            tree_model, tree_iter = selection.get_selected()
+            tree_path = tree_model.get_path(tree_iter)
+            index = tree_path.get_indices()[0]
+
+            handler(widget, row=index)
+
+        self._connections.append(self.table.connect("cursor-changed", _handler))
```

Comment:
The handler can be a `callable` or `None`. I think you don't handle the `None` case. Just check if the `on_select` is not `None`.